nexus/reconfigurator/planning/src/blueprint_builder/external_networking.rs

```diff
@@ -149,6 +178,38 @@ impl<'a> BuilderExternalNetworking<'a> {
             }
         }
 
+        // Recycle the IP addresses of expunged or about-to-be-expunged
+        // external DNS zones. The latter is an unfortunate consequence
+        // of when this builder is constructed. An alternative would be
+        // a hook in expungement, but that seems possibly worse than this
+        // (admittedly awkward) filter.
+        // TODO: Remove when external DNS addresses come from policy.
+        for (_, z) in parent_blueprint
+            .all_omicron_zones(BlueprintZoneFilter::All)
+            .filter(|(sled_id, zone)| {
+                zone.disposition == BlueprintZoneDisposition::Expunged
+                    || input.all_sleds(SledFilter::Commissioned).any(
+                        |(id, sled)| {
+                            id == *sled_id
+                                && sled.policy == SledPolicy::Expunged
+                        },
+                    )
+            })
+        {
+            if let BlueprintZoneType::ExternalDns(dns) = &z.zone_type {
+                match dns.nic.ip {
+                    IpAddr::V4(ip) => existing_external_dns_v4_ips.insert(ip),
+                    IpAddr::V6(ip) => existing_external_dns_v6_ips.insert(ip),
+                };
+                if !available_external_dns_ips.insert(dns.dns_address) {
+                    bail!(
+                        "unavailable external DNS IP: {}",
+                        dns.dns_address.addr
+                    );
+                }
```

Comment:
I'm a little confused by this logic.  It looks like for each expunged zone, we're inserting its IP into the "existing" map.  That seems backwards from what I'd expect since we're treating IPs from expunged zones as available and I thought this map contains _un_available IPs.  But we're also inserting its sockaddr into a map of _available_ ips?

Reply:
Quite right, thank you! I think I was confused for a while about those maps. This can be removed.

Reply:
Fixed in e94c9c024.